### ISSUE TITLE
如何计算投影梯度？

### ISSUE THREAD

**Issue by xgxgnpu** [zxgcqupt/PINNs4SRA#3]
您好！我之前看了您的代码实现和PCGrad原作者的代码，原作者代码如下[PCGrad](https://github.com/tianheyu927/PCGrad/blob/master/PCGrad_tf.py)，对代码实现有一些疑问。

- 1、您这个算例[PINN_for_medium_homo_CTMC_reliability_analysis.ipynb](https://github.com/zxgcqupt/PINNs4SRA/blob/main/medium_homogeneous_CTMC/PINN_for_medium_homo_CTMC_reliability_analysis.ipynb)中，计算梯度向量内积的这行代码inner_product = tf.reduce_sum(grad_task*grads_task[k])中，grad_task是二维张量（里面是两个loss的梯度向量），而grads_task[k]是一维张量（里面是其中一个loss的梯度向量），这样内积再求和，得到的也不是两个向量的内积。原作者的代码实现也有这个问题。请问您是怎么理解这一行代码的呢？

**Comment by zxgcqupt** [comment]
向量的内积就是inner_product = tf.reduce_sum(grad_task*grads_task[k])这样算的。你是不是理解错了？

a = np.random.rand(100)
b = np.random.rand(100)

其中，内积操作np.inner(a, b)和np.sum(a*b)是一样的。

**Comment by xgxgnpu** [comment]
> 

是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）

**Comment by xgxgnpu** [comment]
昨天提到的PCGrad存在训练发散的问题，单独使用PCGrad训练十万次（可能几万次就会发散了），就会出现发散。可以通过调整投影方向来解决，选择用a向量投影到b向量上，还是用b向量投影到a向量上，需要具体问题具体分析的，我和我同事都测试过，您给的算例确实训练发散了，不过可以进行调整优化PCGrad算法来解决的。

**Comment by zxgcqupt** [comment]
> > 
> 
> 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）

你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。

**Comment by xgxgnpu** [comment]
> 

我觉得原作者给的代码确实有些问题，我给原作者发邮件确认一下。您说的函数tf.vectorized_map只是用来做并行加速计算的，跟张量维度没有关系，对于深度学习中的张量tensor的计算，最重要的就是它的维度了，数学定义（比如内积，矩阵乘法）要尊重张量维度才能得到正确的数学含义

**Comment by xgxgnpu** [comment]
> > > 
> > 
> > 
> > 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）
> 
> 你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。

`            grads_task = tf.vectorized_map(lambda x: tf.concat([tf.reshape(grad, [-1,]) 
                                for grad in tape.gradient(x, self.model.trainable_variables)
                                if grad is not None], axis=0), loss)`

这里loss有两个分量，tf.reshape(grad, [-1,])是一维张量，然后tf.concat得到的grads_task是二维张量，在tf.reduce_sum(grad_task*grads_task[k])中，因此grad_task是二维张量，grad_task[k]是一维张量，得到的肯定不是两个梯度向量的内积啊

**Comment by xgxgnpu** [comment]
> > > > 
> > > 
> > > 
> > > 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）
> > 
> > 
> > 你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。
> 
> ` grads_task = tf.vectorized_map(lambda x: tf.concat([tf.reshape(grad, [-1,]) for grad in tape.gradient(x, self.model.trainable_variables) if grad is not None], axis=0), loss)`
> 
> 这里loss有两个分量，tf.reshape(grad, [-1,])是一维张量，然后tf.concat得到的grads_task是二维张量，在tf.reduce_sum(grad_task*grads_task[k])中，因此grad_task是二维张量，grad_task[k]是一维张量，得到的肯定不是两个梯度向量的内积啊

打印出来的张量维度：grads_task的形状： (2, 5350)，grads_task1也就是grads_task[k]的形状： (5350,)
这两个张量维度不一样，所以相乘无法得到真正的内积，最后得到的是内积再加上grads_task[k]的二范数模长
loss的形状： (2,)
loss的len： 2
grads_task的形状： (2, 5350)
grads_task1的形状： (5350,)
grads_task2的形状： (5350,)
新梯度grads_task_new形状： (2, 5350)
新梯度grads_task_new[0]形状： (5350,)
新梯度grads_task_new[1]形状： (5350,)
task1的梯度大小： Tensor("reduce_variance/Mean_1:0", shape=(), dtype=float32)
task2的梯度大小： Tensor("reduce_variance_1/Mean_1:0", shape=(), dtype=float32)
num_tasks数量是： 2
grads_task_new[0]的形状： Tensor("strided_slice_20:0", shape=(5350,), dtype=float32)
inner_product内积形状： ()
proj_direction内积形状： ()
grads_task_newnew0形状： (5350,)
grads_task_newnew1形状： (5350,)
grads_task_newnew形状： (2, 5350)
self.model.trainable_variables列表形状： 7

**Comment by zxgcqupt** [comment]
> > > > > 
> > > > 
> > > > 
> > > > 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）
> > > 
> > > 
> > > 你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。
> > 
> > 
> > ` grads_task = tf.vectorized_map(lambda x: tf.concat([tf.reshape(grad, [-1,]) for grad in tape.gradient(x, self.model.trainable_variables) if grad is not None], axis=0), loss)`
> > 这里loss有两个分量，tf.reshape(grad, [-1,])是一维张量，然后tf.concat得到的grads_task是二维张量，在tf.reduce_sum(grad_task*grads_task[k])中，因此grad_task是二维张量，grad_task[k]是一维张量，得到的肯定不是两个梯度向量的内积啊
> 
> 打印出来的张量维度：grads_task的形状： (2, 5350)，grads_task1也就是grads_task[k]的形状： (5350,) 这两个张量维度不一样，所以相乘无法得到真正的内积，最后得到的是内积再加上grads_task[k]的二范数模长 loss的形状： (2,) loss的len： 2 grads_task的形状： (2, 5350) grads_task1的形状： (5350,) grads_task2的形状： (5350,) 新梯度grads_task_new形状： (2, 5350) 新梯度grads_task_new[0]形状： (5350,) 新梯度grads_task_new[1]形状： (5350,) task1的梯度大小： Tensor("reduce_variance/Mean_1:0", shape=(), dtype=float32) task2的梯度大小： Tensor("reduce_variance_1/Mean_1:0", shape=(), dtype=float32) num_tasks数量是： 2 grads_task_new[0]的形状： Tensor("strided_slice_20:0", shape=(5350,), dtype=float32) inner_product内积形状： () proj_direction内积形状： () grads_task_newnew0形状： (5350,) grads_task_newnew1形状： (5350,) grads_task_newnew形状： (2, 5350) self.model.trainable_variables列表形状： 7

在下边这个函数里，grad_task和grads_task[k]的shape应该是一样的，这样计算inner product没有任何问题。我已经在我这边验证了grad_task和grads_task[k]的shape，它们是一样的。麻烦你再次检查下。
```
def proj_grad(grad_task):
     tf.print ('shape: ', grad_task.shape)
     tf.print ('task k shape: ', grads_task[0].shape)
    
     for k in range(num_tasks):
         inner_product = tf.reduce_sum(grad_task*grads_task[k])
         proj_direction = inner_product / tf.reduce_sum(grads_task[k]*grads_task[k])
         grad_task = grad_task - tf.minimum(proj_direction, 0.) * grads_task[k]
     return grad_task
```

**Comment by xgxgnpu** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）
> > > > 
> > > > 
> > > > 你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。
> > > 
> > > 
> > > ` grads_task = tf.vectorized_map(lambda x: tf.concat([tf.reshape(grad, [-1,]) for grad in tape.gradient(x, self.model.trainable_variables) if grad is not None], axis=0), loss)`
> > > 这里loss有两个分量，tf.reshape(grad, [-1,])是一维张量，然后tf.concat得到的grads_task是二维张量，在tf.reduce_sum(grad_task*grads_task[k])中，因此grad_task是二维张量，grad_task[k]是一维张量，得到的肯定不是两个梯度向量的内积啊
> > 
> > 
> > 打印出来的张量维度：grads_task的形状： (2, 5350)，grads_task1也就是grads_task[k]的形状： (5350,) 这两个张量维度不一样，所以相乘无法得到真正的内积，最后得到的是内积再加上grads_task[k]的二范数模长 loss的形状： (2,) loss的len： 2 grads_task的形状： (2, 5350) grads_task1的形状： (5350,) grads_task2的形状： (5350,) 新梯度grads_task_new形状： (2, 5350) 新梯度grads_task_new[0]形状： (5350,) 新梯度grads_task_new[1]形状： (5350,) task1的梯度大小： Tensor("reduce_variance/Mean_1:0", shape=(), dtype=float32) task2的梯度大小： Tensor("reduce_variance_1/Mean_1:0", shape=(), dtype=float32) num_tasks数量是： 2 grads_task_new[0]的形状： Tensor("strided_slice_20:0", shape=(5350,), dtype=float32) inner_product内积形状： () proj_direction内积形状： () grads_task_newnew0形状： (5350,) grads_task_newnew1形状： (5350,) grads_task_newnew形状： (2, 5350) self.model.trainable_variables列表形状： 7
> 
> 在下边这个函数里，grad_task和grads_task[k]的shape应该是一样的，这样计算inner product没有任何问题。我已经在我这边验证了grad_task和grads_task[k]的shape，它们是一样的。麻烦你再次检查下。
> 
> ```
> def proj_grad(grad_task):
>      tf.print ('shape: ', grad_task.shape)
>      tf.print ('task k shape: ', grads_task[0].shape)
>     
>      for k in range(num_tasks):
>          inner_product = tf.reduce_sum(grad_task*grads_task[k])
>          proj_direction = inner_product / tf.reduce_sum(grads_task[k]*grads_task[k])
>          grad_task = grad_task - tf.minimum(proj_direction, 0.) * grads_task[k]
>      return grad_task
> ```
是的，您打印的这个是对的，我跟您打印的位置不一样，我打印的是函数外部 grads_task的形状。找到原因了，
`
def proj_grad(grad_task):
     tf.print ('shape: ', grad_task.shape)
     tf.print ('task k shape: ', grads_task[0].shape)
    
     for k in range(num_tasks):
         inner_product = tf.reduce_sum(grad_task*grads_task[k])
         proj_direction = inner_product / tf.reduce_sum(grads_task[k]*grads_task[k])
         grad_task = grad_task - tf.minimum(proj_direction, 0.) * grads_task[k]
     return grad_task

#并行计算，grads_task中包含两列，分别传入proj_grad函数进行计算的
proj_grads_flatten = tf.vectorized_map(proj_grad, grads_task)

`

之前我把计算内积的那一行看错了，实际上是grad_task*grads_task[k]，我一直看成了grads_task*grads_task[k]（多了一个s，如果是这个，那就形状不对了）。

之前的问题都是我自己粗心造成的，对给您造成的问题和困扰表示抱歉。

不过，您实现pcgrad时，好像没有进行梯度标准化（一个梯度大，一个梯度小，投影效果不好），所以投影之后也会造成梯度朝着某一个主要方向更新，对于整体学习精度提升不明显，目前的这个pcgrad算法对问题的本身的梯度特征比较敏感。

**Comment by zxgcqupt** [comment]
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > 是的，逐点对应相乘再求和得到内积，这没问题。问题在于grad_task是一个二维张量，维度是（2，3262），但grad_task[k]是一维度张量，维度是（3262，）, 虽然python的广播机制可以让这两个张量计算，但是这个式子tf.reduce_sum(grad_task*grads_task[k])得到的肯定不是两个向量的内积（除了内积之外，应该还多出了向量grads_task[k]的逐点平方和）
> > > > > 
> > > > > 
> > > > > 你理解的好像不对，你可以print 出grad_task和grads_task[k]，还有注意看下tf.vectorized_map这个函数。
> > > > 
> > > > 
> > > > ` grads_task = tf.vectorized_map(lambda x: tf.concat([tf.reshape(grad, [-1,]) for grad in tape.gradient(x, self.model.trainable_variables) if grad is not None], axis=0), loss)`
> > > > 这里loss有两个分量，tf.reshape(grad, [-1,])是一维张量，然后tf.concat得到的grads_task是二维张量，在tf.reduce_sum(grad_task*grads_task[k])中，因此grad_task是二维张量，grad_task[k]是一维张量，得到的肯定不是两个梯度向量的内积啊
> > > 
> > > 
> > > 打印出来的张量维度：grads_task的形状： (2, 5350)，grads_task1也就是grads_task[k]的形状： (5350,) 这两个张量维度不一样，所以相乘无法得到真正的内积，最后得到的是内积再加上grads_task[k]的二范数模长 loss的形状： (2,) loss的len： 2 grads_task的形状： (2, 5350) grads_task1的形状： (5350,) grads_task2的形状： (5350,) 新梯度grads_task_new形状： (2, 5350) 新梯度grads_task_new[0]形状： (5350,) 新梯度grads_task_new[1]形状： (5350,) task1的梯度大小： Tensor("reduce_variance/Mean_1:0", shape=(), dtype=float32) task2的梯度大小： Tensor("reduce_variance_1/Mean_1:0", shape=(), dtype=float32) num_tasks数量是： 2 grads_task_new[0]的形状： Tensor("strided_slice_20:0", shape=(5350,), dtype=float32) inner_product内积形状： () proj_direction内积形状： () grads_task_newnew0形状： (5350,) grads_task_newnew1形状： (5350,) grads_task_newnew形状： (2, 5350) self.model.trainable_variables列表形状： 7
> > 
> > 
> > 在下边这个函数里，grad_task和grads_task[k]的shape应该是一样的，这样计算inner product没有任何问题。我已经在我这边验证了grad_task和grads_task[k]的shape，它们是一样的。麻烦你再次检查下。
> > ```
> > def proj_grad(grad_task):
> >      tf.print ('shape: ', grad_task.shape)
> >      tf.print ('task k shape: ', grads_task[0].shape)
> >     
> >      for k in range(num_tasks):
> >          inner_product = tf.reduce_sum(grad_task*grads_task[k])
> >          proj_direction = inner_product / tf.reduce_sum(grads_task[k]*grads_task[k])
> >          grad_task = grad_task - tf.minimum(proj_direction, 0.) * grads_task[k]
> >      return grad_task
> > ```
> 
> 是的，您打印的这个是对的，我跟您打印的位置不一样，我打印的是函数外部 grads_task的形状。找到原因了， ` def proj_grad(grad_task): tf.print ('shape: ', grad_task.shape) tf.print ('task k shape: ', grads_task[0].shape)
> 
> ```
>  for k in range(num_tasks):
>      inner_product = tf.reduce_sum(grad_task*grads_task[k])
>      proj_direction = inner_product / tf.reduce_sum(grads_task[k]*grads_task[k])
>      grad_task = grad_task - tf.minimum(proj_direction, 0.) * grads_task[k]
>  return grad_task
> ```
> 
> #并行计算，grads_task中包含两列，分别传入proj_grad函数进行计算的 proj_grads_flatten = tf.vectorized_map(proj_grad, grads_task)
> 
> `
> 
> 之前我把计算内积的那一行看错了，实际上是grad_task_grads_task[k]，我一直看成了grads_task_grads_task[k]（多了一个s，如果是这个，那就形状不对了）。
> 
> 之前的问题都是我自己粗心造成的，对给您造成的问题和困扰表示抱歉。
> 
> 不过，您实现pcgrad时，好像没有进行梯度标准化（一个梯度大，一个梯度小，投影效果不好），所以投影之后也会造成梯度朝着某一个主要方向更新，对于整体学习精度提升不明显，目前的这个pcgrad算法对问题的本身的梯度特征比较敏感。

“**不过，您实现pcgrad时，好像没有进行梯度标准化（一个梯度大，一个梯度小，投影效果不好），所以投影之后也会造成梯度朝着某一个主要方向更新，对于整体学习精度提升不明显，目前的这个pcgrad算法对问题的本身的梯度特征比较敏感**”，针对这一问题，你可否发一些相关的论文参考下，或者把论文的链接贴在这里。多谢。
